controls/gql_mutations.py

```diff
@@ -0,0 +1,160 @@
+import logging
+
+import graphene
+from django.contrib.auth.models import AnonymousUser
+from django.core.exceptions import ValidationError, PermissionDenied
+from django.db import transaction
+from graphene import InputObjectType
+
+from contribution.gql_mutations import PremiumBase, update_or_create_premium
+from contribution.models import Premium
+
+from core.schema import OpenIMISMutation
+from insuree.gql_mutations import FamilyBase, InsureeBase
+from insuree.services import FamilyService, InsureeService
+from controls.models import MobileEnrollmentMutation as MobileMutationLog
+from policy.gql_mutations import PolicyInputType
+from policy.models import Policy
+from policy.services import PolicyService
+from controls.apps import ControlsConfig
+
+
+logger = logging.getLogger(__name__)
+
+
+class PremiumEnrollmentGQLType(PremiumBase, InputObjectType):
+    policy_id = graphene.Int(required=True)
+
+
+class PolicyEnrollmentGQLType(PolicyInputType, InputObjectType):
+    mobile_id = graphene.Int(required=True)
+
+
+class InsureeEnrollmentGQLType(InsureeBase, InputObjectType):
+    pass
+
+
+class FamilyEnrollmentGQLType(FamilyBase, InputObjectType):
+    pass
+
+
+class MobileEnrollmentGQLType:
+    family = graphene.Field(FamilyEnrollmentGQLType, required=True)
+    insurees = graphene.List(InsureeEnrollmentGQLType)  # for families with more than the head insuree
+    policies = graphene.List(PolicyEnrollmentGQLType, required=True)
+    premiums = graphene.List(PremiumEnrollmentGQLType, required=True)
+
+
+MOBILE_ENROLLMENT_RIGHTS = [
+    ControlsConfig.gql_mutation_create_families_perms,
+    ControlsConfig.gql_mutation_update_families_perms,
+    ControlsConfig.gql_mutation_create_insurees_perms,
+    ControlsConfig.gql_mutation_update_insurees_perms,
+    ControlsConfig.gql_mutation_create_policies_perms,
+    ControlsConfig.gql_mutation_edit_policies_perms,
+    ControlsConfig.gql_mutation_create_premiums_perms,
+    ControlsConfig.gql_mutation_update_premiums_perms,
+]
+
+
+class MobileEnrollmentMutation(OpenIMISMutation):
+    _mutation_module = "controls"
+    _mutation_class = "MobileEnrollmentMutation"
+
+    class Input(MobileEnrollmentGQLType, OpenIMISMutation.Input):
+        pass
+
+    @classmethod
+    def async_mutate(cls, user, **data):
+        logger.info("Receiving new mobile enrollment request")
+        logger.info(data)
+        try:
+            if type(user) is AnonymousUser or not user.id:
+                raise ValidationError("mutation.authentication_required")
+            if any(not user.has_perms(right) for right in MOBILE_ENROLLMENT_RIGHTS):
+                raise PermissionDenied("unauthorized")
```

Comment:
I was wondering if we should let the specific services to check this. What happens if there are errors (e.g. existing insuree id) in the data with the already saved data ?

Reply:
I think validation is already handled by the service calls in lines 92 and 99. My guess is they added this here as an early return, canceling the operation early if the user does not have any of the required (similar to validate_mutation we have in newer modules)